src/include/detail/linalg/tdb_matrix_with_ids.h

```diff
@@ -0,0 +1,272 @@
+/**
+ * @file   tdb_matrix_with_ids.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Class that provides a matrix interface to two TileDB arrays, one containing
+ * vectors and one IDs.
+ *
+ * @todo Include the right headers for BLAS support.
+ * @todo Refactor ala tdb_partitioned_matrix.h
+ *
+ */
+
+#ifndef TDB_MATRIX_WITH_IDS_H
+#define TDB_MATRIX_WITH_IDS_H
+
+#include "detail/linalg/matrix_with_ids.h"
+#include "detail/linalg/tdb_helpers.h"
+#include "detail/linalg/tdb_matrix.h"
+#include "tdb_defs.h"
+
+/**
+ * Derived from `tdbBlockedMatrix`, which we have derive from `MatrixWithIds`.
+ * Initialized in construction by filling from the TileDB vectors array and the
+ * the TileDB IDs array.
+ */
+template <
+    class T,
+    class LayoutPolicy = stdx::layout_right,
+    class I = size_t,
+    class IdsType = size_t>
+class tdbBlockedMatrixWithIds
+    : public tdbBlockedMatrix<
+          T,
+          LayoutPolicy,
+          I,
+          MatrixWithIds<T, LayoutPolicy, I, IdsType>> {
+  using Base = tdbBlockedMatrix<
+      T,
+      LayoutPolicy,
+      I,
+      MatrixWithIds<T, LayoutPolicy, I, IdsType>>;
+  using Base::Base;
+
```

Comment:
As with `Matrix` and `MatrixWithIds`, `tdbMatrixWithIds` needs to own the ids.  The inheritance hierarchy should be the same and the ids belong in the "WithIds" class.